src/main/java/org/springframework/data/couchbase/core/convert/join/N1qlJoinResolver.java

```diff
@@ -94,38 +109,63 @@ public static String buildQuery(CouchbaseTemplate template, N1qlJoinResolverPara
 		return statementSb.toString();
 	}
 
-	public static <R> List<R> doResolve(CouchbaseTemplate template, N1qlJoinResolverParameters parameters,
-			Class<R> associatedEntityClass) {
-		throw new UnsupportedOperationException();
-		/*
-		    String statement = buildQuery(template, parameters);
-		
-		    if (LOGGER.isDebugEnabled()) {
-		        LOGGER.debug("Join query executed " + statement);
-		    }
-		
-		    N1QLQuery query = new N1QLQuery(N1QLExpression.x(statement), QueryOptions.queryOptions());
-		    return template.findByN1QL(query, associatedEntityClass);*/
+	public static <R> List<R> doResolve(CouchbaseTemplate template, String collectionName,
+			N1qlJoinResolverParameters parameters, Class<R> associatedEntityClass) {
+
+		String statement = buildQuery(template.reactive(), collectionName, parameters);
+
+		if (LOGGER.isDebugEnabled()) {
+			LOGGER.debug("Join query executed " + statement);
+		}
+
+		N1QLQuery query = new N1QLQuery(N1QLExpression.x(statement), QueryOptions.queryOptions());
+		List<R> result = template.findByQuery(associatedEntityClass).matching(query).all();
+		return result.isEmpty() ? null : result;
+	}
+
+	public static <R> List<R> doResolve(ReactiveCouchbaseTemplate template, String collectionName,
+			N1qlJoinResolverParameters parameters, Class<R> associatedEntityClass) {
+
+		String statement = buildQuery(template, collectionName, parameters);
+
+		if (LOGGER.isDebugEnabled()) {
+			LOGGER.debug("Join query executed " + statement);
+		}
+
+		N1QLQuery query = new N1QLQuery(N1QLExpression.x(statement), QueryOptions.queryOptions());
+		List<R> result = template.findByQuery(associatedEntityClass).matching(query).all().collectList().block();
+		return result.isEmpty() ? null : result;
 	}
 
 	public static boolean isLazyJoin(N1qlJoin joinDefinition) {
 		return joinDefinition.fetchType().equals(FetchType.LAZY);
 	}
 
 	static public class N1qlJoinProxy implements InvocationHandler {
-		private final CouchbaseTemplate template;
+		private final Object template;
+		private final String collectionName = null;
 		private final N1qlJoinResolverParameters params;
 		private List<?> resolved = null;
 
-		public N1qlJoinProxy(CouchbaseTemplate template, N1qlJoinResolverParameters params) {
+		public N1qlJoinProxy(Object template, N1qlJoinResolverParameters params) {
```

Comment:
the other option is to make the constructor accept both the regular and the reactive one and the caller just nulls out the one it doesn't use.. that way it still keeps the type safety.

Reply:
I made two constructors. Fixed.